### ISSUE TITLE
Question about default_capabilities

### ISSUE THREAD

**Issue by weeman1337** [hrsh7th/cmp-nvim-lsp#44]
Hello :wave: Just upgraded my setup and found the change about client capabilities.

All examples look like this:

```
-- The nvim-cmp almost supports LSP's capabilities so You should advertise it to LSP servers..
local capabilities = require('cmp_nvim_lsp').default_capabilities()

-- The following example advertise capabilities to `clangd`.
require'lspconfig'.clangd.setup {
  capabilities = capabilities,
}
```

Doesn't that mean that I not consider all other capabilities, that nvim has?

Maybe something like this would be better:

```
local capabilities = vim.lsp.protocol.make_client_capabilities()
local cmp_capabilities = require("cmp_nvim_lsp").default_capabilities()
capabilities.textDocument.completion = cmp_capabilities.textDocument.completion
```

**Comment by skwee357** [comment]
Following. I'm also not sure how to migrate properly, because before the breaking change, I was using

```
    local capabilities = vim.lsp.protocol.make_client_capabilities()
    capabilities = require('cmp_nvim_lsp').update_capabilities(capabilities)
```

**Comment by efueyo** [comment]
> Following. I'm also not sure how to migrate properly, because before the breaking change, I was using
> 
> ```
>     local capabilities = vim.lsp.protocol.make_client_capabilities()
>     capabilities = require('cmp_nvim_lsp').update_capabilities(capabilities)
> ```

I just followed what was commented on https://github.com/hrsh7th/cmp-nvim-lsp/issues/38#issuecomment-1279744539
and it seemed to work for me @skwee357

**Comment by dimbleby** [comment]
https://github.com/hrsh7th/cmp-nvim-lsp/issues/42#issuecomment-1283825572 looks like better advice.

The suggestion at the previous comment - and currently in the README - goes wrong exactly as this issue reports: it throws away all the capabilities declared by the the core neovim client.

**Comment by mgnsk** [comment]
> [#42 (comment)](https://github.com/hrsh7th/cmp-nvim-lsp/issues/42#issuecomment-1283825572) looks like better advice.
> 
> The suggestion at the previous comment - and currently in the README - goes wrong exactly as this issue reports: it throws away all the capabilities declared by the the core neovim client.

After some research I can confirm that this works at the moment. Extending the default capabilities with the specified new capabilities.
```lua
            local lsp = require("lspconfig")
            lsp.util.default_config = vim.tbl_deep_extend("force", lsp.util.default_config, {
                capabilities = require("cmp_nvim_lsp").default_capabilities(),
            })
```

But https://github.com/hrsh7th/nvim-cmp/issues/1265 is open, I think it also needs documentation in README as well.

**Comment by liskin** [comment]
I don't think this is a problem. nvim-lspconfig internally performs `tbl_deep_extend` to [merge nvim's capabilities with defaults for a given language server](https://github.com/neovim/nvim-lspconfig/blob/511609ae0311abfcfaed3c398429a147e895ce2c/lua/lspconfig/configs.lua#L60) and then again to [merge that with the user configuration (cmp_nvim_lsp's default_capabilities)](https://github.com/neovim/nvim-lspconfig/blob/511609ae0311abfcfaed3c398429a147e895ce2c/lua/lspconfig/configs.lua#L90). The examples work correctly indeed, and I've verified by doing `print(vim.inspect(config.capabilities))` in on_setup that the resulting capabilities structure has everything it should have.

**Comment by charypar** [comment]
This seems like the closes issue to the root cause of a problem I've spent a day and night chasing down. 

The symptom was rust-analyzer hanging for a long time at the startup in a large directory, saying "Roots Scanned". At the same time `fseventsd` on macOS was using close to a 100% of a CPU core, while rust-analyzer was hovering around 18%.

Not sure of the specific interaction of plugins, but using `require("cmp_nvim_lsp").default_capabilities()` with mason lspconfig seems to override nvim lsp's defaults. 

In my case, the key setting was `workspace.didChangeWatchedFiles.dynamicRegistration` which needs to be set to `true` for rust-analyzer to work reasonably fast in large directories (my hypothesis is that when set to false, rust analyzer attempts to set up a watch with the OS on _every_ file in the workspace). 

As far as I can tell, the default [is](https://github.com/neovim/neovim/blob/bb032d952bfc692062fceb764ff2742c5bdd3324/runtime/lua/vim/lsp/protocol.lua#L818-L844) `true`, but when setting up the LSP like so:

```lua
        local lspconfig = require("lspconfig")
        local mason_lspconfig = require("mason-lspconfig")
        local cmp_nvim_lsp = require("cmp_nvim_lsp")

        local capabilities = cmp_nvim_lsp.default_capabilities()

        mason_lspconfig.setup_handlers({
            function(server_name)
                lspconfig[server_name].setup({
                    capabilities = capabilities,
                })
            end,
        })
```

it ends up being set to `false` - rust-analyzer behaves differently (and correctly) if I add this line:

```lua
        local capabilities = cmp_nvim_lsp.default_capabilities()
        capabilities.workspace = { didChangeWatchedFiles = { dynamicRegistration = true } } -- <- addded line
```

Not sure if `mason-lspconfig` can influence this, as I read the code, it simply calls the handler, which talks to `lspconfig` directly.

This is as far as I got, being new to neovim, I'm not sure how to trace and debug the configuration values for definitive proof, but hopefully it helps.
